aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cu

```diff
@@ -14,11 +14,89 @@
+  cublasFillMode_t uplo = upper ? CUBLAS_FILL_MODE_UPPER : CUBLAS_FILL_MODE_LOWER;
+  cublasOperation_t trans = transpose ? CUBLAS_OP_T : CUBLAS_OP_N;
+  trans = conjugate_transpose ? CUBLAS_OP_C : trans;
+  cublasDiagType_t diag = unitriangular ? CUBLAS_DIAG_UNIT : CUBLAS_DIAG_NON_UNIT;
+  cublasSideMode_t side = CUBLAS_SIDE_LEFT;
+
+  auto A_data = A.data_ptr<scalar_t>();
+  auto B_data = B.data_ptr<scalar_t>();
+  auto A_mat_stride = matrixStride(A);
+  auto B_mat_stride = matrixStride(B);
+  auto batch_size = batchCount(A);
+  auto n = cuda_int_cast(A.size(-2), "n");
+  auto nrhs = cuda_int_cast(B.size(-1), "nrhs");
+  auto lda = std::max<int>(1, n);
+
+  auto alpha = scalar_t{1};
+
+  for (decltype(batch_size) i = 0; i < batch_size; i++) {
+    scalar_t* A_working_ptr = &A_data[i * A_mat_stride];
+    scalar_t* B_working_ptr = &B_data[i * B_mat_stride];
+    auto handle = at::cuda::getCurrentCUDABlasHandle();
+    at::cuda::blas::trsm(handle, side, uplo, trans, diag, n, nrhs, &alpha, A_working_ptr, lda, B_working_ptr, lda);
+  }
+}
+
+void triangular_solve_cublas(Tensor& A, Tensor& B, Tensor& infos, bool upper, bool transpose, bool conjugate_transpose, bool unitriangular) {
+  (void)infos; // unused
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(A.scalar_type(), "triangular_solve_cuda", [&]{
+    apply_triangular_solve<scalar_t>(A, B, upper, transpose, conjugate_transpose, unitriangular);
+  });
+}
+
+template <typename scalar_t>
+static void apply_triangular_solve_batched(Tensor& A, Tensor& B, bool upper, bool transpose, bool conjugate_transpose, bool unitriangular) {
+  cublasFillMode_t uplo = upper ? CUBLAS_FILL_MODE_UPPER : CUBLAS_FILL_MODE_LOWER;
+  cublasOperation_t trans = transpose ? CUBLAS_OP_T : CUBLAS_OP_N;
+  trans = conjugate_transpose ? CUBLAS_OP_C : trans;
+  cublasDiagType_t diag = unitriangular ? CUBLAS_DIAG_UNIT : CUBLAS_DIAG_NON_UNIT;
+  cublasSideMode_t side = CUBLAS_SIDE_LEFT;
+
+  auto A_data = A.data_ptr<scalar_t>();
+  auto B_data = B.data_ptr<scalar_t>();
+  auto A_mat_stride = matrixStride(A);
+  auto B_mat_stride = matrixStride(B);
+  auto batch_size = cuda_int_cast(batchCount(A), "batch_size");
+  auto n = cuda_int_cast(A.size(-2), "n");
+  auto nrhs = cuda_int_cast(B.size(-1), "nrhs");
+  auto lda = std::max<int>(1, n);
+
+  auto alpha = scalar_t{1};
+
+  // cuBLAS batched trsm requires input to be the device array of pointers to device single matrices
+  Tensor A_array = at::arange(
+    reinterpret_cast<int64_t>(A_data),
+    reinterpret_cast<int64_t>(&A_data[(std::max<int>(batch_size, 1) - 1) * A_mat_stride]) + 1,
+    static_cast<int64_t>(std::max<int64_t>(A_mat_stride, 1) * sizeof(scalar_t)), A.options().dtype(at::kLong));
+
+  Tensor B_array = at::arange(
+    reinterpret_cast<int64_t>(B_data),
+    reinterpret_cast<int64_t>(&B_data[(std::max<int>(batch_size, 1) - 1) * B_mat_stride]) + 1,
+    static_cast<int64_t>(std::max<int64_t>(B_mat_stride, 1) * sizeof(scalar_t)), B.options().dtype(at::kLong));
```

Comment:
I assume these `max(..., 1)` statements are used for zero sized or zero batched matrix? Do we have those tests to make sure they don't cause cuda illegal memory access or other `arange` errors? Can you also check if a cuda kernel is launched, or cuda device/stream synchronization is called in this case? Those could negatively affect the performance.

Besides that, I still think it's not good to handle zero-numel input cases in such deep locations and dependent on library implementations. See discussions https://github.com/pytorch/pytorch/pull/53104#discussion_r586829940. 

cc @mruberry @heitorschueroff 

cc @ngimel in case interested

Reply:
Inside `trsmBatched` no cuda kernel and no synchronization is called if the batch size is zero, cuda-memcheck is also clear.

All zero-batched tensors use null pointer:
```python
In [4]: import torch
In [5]: a = torch.randn(2, 0, 2, 2, device='cuda')
In [6]: a1 = torch.randn(3, 0, 2, 2, device='cuda')
In [7]: a.data_ptr()
Out[7]: 0
In [8]: a1.data_ptr()
Out[8]: 0
In [9]: a[1].data_ptr()
Out[9]: 16
In [10]: a1[1].data_ptr()
Out[10]: 16
```
So a device tensor of pointers to matrices `arange(start=0, end=1, step=sizeof(dtype)*matrix_stride)` would be correct here, though not referenced anyway.

I understand your concern. But this is not a special case as it is a perfectly valid input for LAPACK/cuBLAS/cuSOLVER that is documented. An error is given only if matrix dimensions or batch size < 0.

The case when the matrix is 1x1 is also trivial, yet we don't make a distinction in the implementation.

Reply:
I can appreciate @xwang233's point for readability, and I have a distrust of math libraries. That said, if this works that's great. How about we split the difference here and add a comment noting this case is correctly handled by the math library? It's not obvious it would be.